components/Collections/src/private/MDCCollectionViewStyler.m

```diff
@@ -246,6 +246,35 @@ - (void)setCellStyle:(MDCCollectionViewCellStyle)cellStyle {
   _cellStyle = cellStyle;
 }
 
+- (BOOL)shouldHideSeparatorForCellLayoutAttributes:(MDCCollectionViewLayoutAttributes *)attr {
+  BOOL shouldHideSeparator = self.shouldHideSeparators;
+  if (!self.delegate)
+    return shouldHideSeparator;
+
+  NSIndexPath *indexPath = attr.indexPath;
+  BOOL isSectionHeader =
+      [attr.representedElementKind isEqualToString:UICollectionElementKindSectionHeader];
+  BOOL isSectionFooter =
+      [attr.representedElementKind isEqualToString:UICollectionElementKindSectionFooter];
+  BOOL isCell = attr.representedElementCategory == UICollectionElementCategoryCell;
+  if (isSectionHeader && [self.delegate respondsToSelector:@selector
+                                        (collectionView:shouldHideHeaderSeparatorForSection:)]) {
+    shouldHideSeparator = [self.delegate collectionView:_collectionView
+                    shouldHideHeaderSeparatorForSection:indexPath.section];
+  }
+  if (isSectionFooter && [self.delegate respondsToSelector:@selector
```

Comment:
Can the same indexPath be a header/footer/cell simultaneously?  If not, it may be better to if/else-if on the cell type first, then have nested conditionals for `respondsToSelector:` inside each.  Placing the `isCell` condition first seems like the primary case and so might be incrementally faster.